gdal/frmts/wms/gdalwmscache.cpp

```diff
@@ -31,89 +33,269 @@
+CPLErr GDALWMSCache::Initialize(const char *pszUrl, CPLXMLNode *pConfig) {
+    const char *pszXmlCachePath = CPLGetXMLValue( pConfig, "Path", nullptr );
+    const char *pszUserCachePath = CPLGetConfigOption( "GDAL_DEFAULT_WMS_CACHE_PATH",
+                                                     nullptr );
+    if( pszXmlCachePath != nullptr )
+    {
+        m_osCachePath = pszXmlCachePath;
+    }
+    else if( pszUserCachePath != nullptr )
+    {
+        m_osCachePath = pszUserCachePath;
+    }
+
+    // Separate folder for each unique dataset url
+    if( CPLTestBool( CPLGetXMLValue( pConfig, "Unique", "True" ) ) )
+    {
+        m_osCachePath = CPLFormFilename( m_osCachePath, CPLMD5String( pszUrl ), nullptr );
+    }
+
+    // TODO: Add sqlite db cache type
+    const char *pszType = CPLGetXMLValue( pConfig, "Type", "file" );
+    if( EQUAL(pszType, "file") )
+    {
+        m_poCache = new FileCache(m_osCachePath, pConfig);
+    }
+
     return CE_None;
 }
 
-CPLErr GDALWMSCache::Read(const char *key, CPLString *file_name) {
-    CPLErr ret = CE_Failure;
-    CPLString cache_file(KeyToCacheFile(key));
-    VSILFILE* fp = VSIFOpenL(cache_file.c_str(), "rb");
-    if (fp != nullptr)
+CPLErr GDALWMSCache::Insert(const char *pszKey, const CPLString &soFileName)
+{
+    if( m_poCache != nullptr && pszKey != nullptr )
     {
-        VSIFCloseL(fp);
-        *file_name = cache_file;
-        ret = CE_None;
+        // Add file to cache
+        CPLErr result = m_poCache->Insert(pszKey, soFileName);
+        if( result == CE_None )
+        {
+            // Start clean thread
+            if( m_hCleanThread == nullptr )
+            {
+                m_hCleanThread = CPLCreateJoinableThread(CleanCacheThread, this);
+            }
+        }
+        return result;
     }
 
-    return ret;
+    return CE_Failure;
 }
 
-CPLString GDALWMSCache::KeyToCacheFile(const char *key) {
-    CPLString hash(MD5String(key));
-    CPLString cache_file(m_cache_path);
+enum GDALWMSCacheItemStatus GDALWMSCache::GetItemStatus(const char *pszKey) const
+{
+    if( m_poCache != nullptr )
+    {
+        return m_poCache->GetItemStatus(pszKey);
+    }
+    return CACHE_ITEM_NOT_FOUND;
+}
 
-    if (!cache_file.empty() && cache_file.back() != '/')
-        cache_file.append(1, '/');
-    for (int i = 0; i < m_cache_depth; ++i) {
-        cache_file.append(1, hash[i]);
-        cache_file.append(1, '/');
+GDALDataset* GDALWMSCache::GetDataset(const char *pszKey,
+                                      char **papszOpenOptions) const
+{
+    if( m_poCache != nullptr )
+    {
+        return m_poCache->GetDataset(pszKey, papszOpenOptions);
     }
-    cache_file.append(hash);
-    cache_file.append(m_postfix);
-    return cache_file;
+    return nullptr;
+}
+
+void GDALWMSCache::Clean()
+{
+    if( m_poCache != nullptr )
+    {
+        CPLDebug("WMS", "Clean cache");
+        m_poCache->Clean();
+    }
+
+    m_hCleanThread = nullptr;
```

Comment:
I don't think m_hCleanThread should be nullified. Because then CPLJoinThread() will not be attempted, and thus the resources associated with the thread will not be cleaned (memory leaks)

Reply:
Fixed